sql/analyzer/inserts.go

```diff
@@ -151,20 +153,46 @@ func wrapRowSource(ctx *sql.Context, insertSource sql.Node, destTbl sql.Table, s
 				}
 			})
 			if err != nil {
-				return nil, false, err
+				return nil, -1, err
 			}
 			projExprs[i] = def
+		} else {
+			projExprs[i] = expression.NewGetField(colIdx, col.Type, col.Name, col.Nullable)
 		}
 
-		if f.AutoIncrement {
+		if col.AutoIncrement {
+			// Regardless of whether the column was explicitly specified, if it is an auto increment column, we need to
+			// wrap it in an AutoIncrement expression.
 			ai, err := expression.NewAutoIncrement(ctx, destTbl, projExprs[i])
 			if err != nil {
-				return nil, false, err
+				return nil, -1, err
 			}
 			projExprs[i] = ai
 
-			if !columnExplicitlySpecified {
-				autoAutoIncrement = true
+			if colIdx == -1 {
+				// Not Explicitly Specified AutoIncrement Column, increment the autoAutoIncrement index immediately
+				autoAutoIncrement = 0
+			} else {
+				// Additionally, the first NULL, DEFAULT, or empty value is what the last_insert_id should be set to.
+				switch src := insertSource.(type) {
+				case *plan.Values:
+					for ii, tup := range src.ExpressionTuples {
+						expr := tup[colIdx]
+						if unwrap, ok := expr.(*expression.Wrapper); ok {
+							expr = unwrap.Unwrap()
+						}
+						if _, isDef := expr.(*sql.ColumnDefaultValue); isDef {
+							autoAutoIncrement = ii
```

Comment:
Ahh... I had to get all the way down here in the code before I understood what the `autoAutoIncrement` variable was representing. How about naming it something like `firstGeneratedAutoIncrementRowIdx`? That's the most self-documenting name I can think of. When I saw `Idx` in another part of this review, I immediately thought of column index, since that's almost always what we mean we we use `idx`, so I think `RowIdx` helps describe what's being represented.